### PR TITLE
feat: Autoscale Delicense feature

### PR DESCRIPTION
**Pull request by ancoleman**
## Description

When utilizing Google Cloud Autoscaling (Managed Instance Groups with Autoscaler Enabled), automatically de-register a firewall with the Panorama Software License Plugin is needed. This de-registration allows the customer to reallocate licenses to their Software NGFW Credit Deployment Profile(s).

## Motivation and Context

The drive to support the automatic de-registration of firewalls within Panorama, is to help minimize the operational overhead of knowing when to de-register an abandoned firewall and which firewall to de-register. This module builds all the necessary components to support the de-registration lifecycle of the firewall.

How does this work?
- Generates a Log Sink with Filter to capture the delete operation from a Managed Instance Group Autoscaler
- Log Sink destination is a pub/sub topic which is also generated by this module
- When the log event happens, it messages the topic which a subscription retrieves the message from the topic
- When the subscription retrieves the message, the trigger subscription connected to a Cloud Function is pushed to the CFN
- The CFN consumes the log event with the instance name/instance id and other relevant information.
- The CFN performs an instance details lookup to retrieve the network interfaces associated with the instance
- The CFN retrieves Panorama credentials from Google Secret Manager
- The CFN uses credentials to connect to Panorama over a VPC Connector (the VPC connector is built by the module)
- The CFN queries the Panorama Plugin to review what firewalls are associated with the plugin and matches the mgmt IP of the firewall with what the CFN queried from Google Cloud. If there is a match, a de-register operation is performed against the specific serial number of the firewall, and then a commit is performed.

Other Items Built by the Module:
- CFN service account with necessary roles
- CFN source code package
- PUB/SUB Topic and Subscriptions
- VPC Connector
- LOG Sink

## How Has This Been Tested?

This was tested locally on MacOS Ventura using Terraform Version 1.3.9. Pre-commit was performed and all checks passed. An exclusion was necessary for Checkov as there is a false positive on not securing the CFN properly. However, the false positive is due to HTTP trigger not being set to SECURE_ALWAYS. HTTP Trigger is irrelevant for this CFN as it is event-driven with Pub/Sub instead. Checkov should revise their rule to enforce ONLY if http_trigger = true or if event_trigger is present to ignore.

## Types of changes

- New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes if appropriate.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by ancoleman** [comment]
not sure why the validate fails, as I most definitely have this functioning with Terraform 1.3.9

**Comment by pavelrn** [comment]
Delicensing code has been merged into `autoscale` example with some minor changes:
- Extra resources are created with TF
  - Storage bucket for Cloud Function code
  - Secret Manager secret
- Cloud Function upgraded to v2 (there are more regions with v2 support then with v1 support)
-  description (README) added

**Review by migara — APPROVED** [review]
@pavelrn great work! Please add a section to the README about licensing/delicensing options before merging

**Review by pimielowski — CHANGES_REQUESTED** [review]
Good job! I found some thing need to be changed, especially in Python code.

File: modules/autoscale/src/main.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    details = get_instance_details(project_id, zone, instance_id)
+    interfaces = details['networkInterfaces']
+    items = details['metadata']['items']
+    igm_name = find_igm_name(items)
+    primary_ip = get_primary_ip(interfaces)
+    info(f"-> Start delicensing process for instance: {instance_id}, primary IP: {primary_ip}")
+
+    # get credentials from Secret Manager and login to Panorama
+    credentials = get_secret()
+    panorama = Panorama(init_parameters().panorama_address, credentials.get('user'), credentials.get('pass'))
```
Comment (sebastianczech) [inline modules/autoscale/src/main.py:40]: Most of the customers are using Panorama working in HA mode, so I propose to extend logic in similar way as we have for AWS, where Lambda for delicensing can work with:
- single Panorama
- Panorama working in HA mode (primary-secondary)

Regarding second case - we need to check which Panorama is active and then do delicensing only on active instance. Feel free to reuse code from AWS: https://github.com/PaloAltoNetworks/terraform-aws-vmseries-modules/blob/e1764506312062be626631d0cd770c918903d9da/modules/asg/scripts/lambda.py#L467-L480

**Review by sebastianczech — CHANGES_REQUESTED** [review]
Please add support for Panorama working in HA mode

File: modules/autoscale/src/main.py (hunk 1 above)
Comment (pavelrn) [reply modules/autoscale/src/main.py]: The delicensing Python code is aligned with the code for the other clouds. Now the code supports Panorama HA

File: modules/autoscale/src/main.py [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+        """
+        secret_name = getenv('SECRET_NAME')
+
+        client = secretmanager.SecretManagerServiceClient()
+        name = f"projects/{self.project_id}/secrets/{secret_name}/versions/latest"
+        response = client.access_secret_version(name=name)
+        secret_str = response.payload.data.decode('UTF-8')
+        
+        self.panorama_username = secret_str.split("|", 1)[0]
+        self.panorama_password = secret_str.split("|", 1)[1]
```
Comment (pimielowski) [inline modules/autoscale/src/main.py:48]: Variables used within a class should be instantiate in init function, you can achieve it like this ->
In line **30** add code `self.panorama_username, self.panorama_password = self.init_panorama_credentials()`
and at line **47** add code `return secret_str.split("|", 1)[0], secret_str.split("|", 1)[1]`

**Review by pimielowski — COMMENTED** [review]
Huge improvement! 
I found some minor stuff only :)

File: modules/autoscale/src/main.py [hunk 3]
```diff
@@ -0,0 +1,264 @@
+import googleapiclient.discovery
+from google.cloud import secretmanager
+import functions_framework
```
[the remark this answers is not in the document]
Comment (pavelrn) [reply modules/autoscale/src/main.py]: Fixed

File: modules/autoscale/src/main.py (hunk 2 above)
Comment (pavelrn) [reply modules/autoscale/src/main.py]: Fixed

File: modules/autoscale/src/main.py [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+from cloudevents.http import CloudEvent
+
+from json import loads
+from logging import getLogger, basicConfig, INFO, DEBUG
+from xml.etree import ElementTree as et
+from xml.etree.ElementTree import Element
+from os import getenv
+import base64
```
[the remark this answers is not in the document]
Comment (pavelrn) [reply modules/autoscale/src/main.py]: Fixed

**Review by pimielowski — APPROVED** [review]
LGTM! 👍

**Pull request merged**